### ISSUE TITLE
[BUG] Service Bus receiving of messages hangs when using processor / sessionProcessor with prefetch

### ISSUE THREAD

**Issue by Maxeh** [Azure/azure-sdk-for-java#37158]
**Describe the bug**
We are using a sessionProcessor to receive messages and when the property prefetchCount is used, the Service Bus randomly hangs and stops processing after some time for sessionIdleTimeout seconds (in the example below 50 seconds). It starts again after the sessionIdleTimeout, but then multiple exceptions occur when complete() is called (reactor.core.Exceptions$ReactiveException: java.lang.InterruptedException).

Without prefetchCount processing is too slow. It seems that this error does not occur when a smaller value < 5 for prefetchCount is used or it happens not that frequently.

This is the configuration of the ServiceBusProcessorClient we use:

```
final ServiceBusProcessorClient receiver = new ServiceBusClientBuilder()
          .credential(namespace, new ClientSecretCredentialBuilder()
                .tenantId(tenantId)
                .clientId(clientId)
                .clientSecret(clientSecret)
                .build())
          .sessionProcessor()
          .disableAutoComplete()
          .receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
          .topicName(topicName)
          .prefetchCount(20)
          .maxConcurrentCalls(1)
          .maxConcurrentSessions(1)
          .sessionIdleTimeout(Duration.ofSeconds(50))
          .subscriptionName(subscriptionName)
          .processError(this::processError)
          .processMessage(this::processMessage)
          .buildProcessorClient();
receiver.start();
```

It seems that this ALSO occurs when using a normal processor without session and with config prefetchCount=10 and maxConcurrentCalls=100. When setting prefetchCount=0 everything works fine. So somehow the prefetch is causing interruptions. The following code also hangs and throws exceptions: 

```
final ServiceBusProcessorClient receiver = new ServiceBusClientBuilder()
          .credential(namespace, new ClientSecretCredentialBuilder()
                .tenantId(tenantId)
                .clientId(clientId)
                .clientSecret(clientSecret)
                .build())
          .processor()
          .disableAutoComplete()
          .receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
          .topicName(topicName)
          .prefetchCount(10)
          .maxConcurrentCalls(100)
          .subscriptionName(subscriptionName)
          .processError(this::processError)
          .processMessage(this::processMessage)
          .buildProcessorClient();
receiver.start();
```

Might be related to: https://github.com/Azure/azure-sdk-for-java/issues/37160

**To Reproduce**
Use the code above to create a ServiceBusProcessorClient, try to receive 1000 messages all with same sessionId. Processing will stop in between and InterruptedExceptions are thrown when complete() is called.

**Expected behavior**
Processing should work without interruptions and exceptions.

**Setup (please complete the following information):**
 - OS: Mac
 - IDE: IntelliJ
 - Library/Libraries: 'com.azure:azure-messaging-servicebus:7.14.4'
 - Java version: 17
 - App Server/Environment: Tomcat
 - Frameworks: Spring Boot

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by joshfree** [comment]
related https://github.com/Azure/azure-sdk-for-java/issues/37160

**Comment by Maxeh** [comment]
Seems to be related to https://github.com/Azure/azure-sdk-for-java/issues/31356 but the proposed fix in this ticket is not working. Smaller values for prefetchCount are working, but when a certain threshold is reached the application hangs. Also it seems that it is connected with maxConcurrentCalls as described in 31356. On my local computer prefetchCount=5 and maxConcurrentCalls=50 is working, but prefetchCount=5 and maxConcurrentCalls=100 causes the app to hang.

**Comment by anuchandy** [comment]
Hello @Maxeh, let me make sure I’ve the setup matching your environment where this problem happens, let’s focus on session-processor.

### Processor Configurations [confirm this]

Prefetch == 20
maxConcurrentCalls == 1
maxConcurrentSessions == 1
sessionIdleTimeout == 50 seconds

### Code [confirm this]

```java
final ServiceBusProcessorClient receiver = new ServiceBusClientBuilder()
          .credential(namespace, new ClientSecretCredentialBuilder()
                .tenantId(tenantId)
                .clientId(clientId)
                .clientSecret(clientSecret)
                .build())
          .sessionProcessor()
          .disableAutoComplete()
          .receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
          .topicName(topicName)
          .prefetchCount(20)
          .maxConcurrentCalls(1)
          .maxConcurrentSessions(1)
          .sessionIdleTimeout(Duration.ofSeconds(50))
          .subscriptionName(subscriptionName)
          .processError(this::processError)
          .processMessage(this::processMessage)
          .buildProcessorClient();
receiver.start();
```

### Additional information needed from you

Please share the following information / clarifications -

1. provide the code showing logic in the `processMessage`, I'm trying to see how much time one invocation of `processMessage` takes before it calls complete.
2. provide the code shwoing logic in the `processError`
3. Is it sufficient to put the above code in a console app (inside static void main() function)?
4. Is there any code after after receiver.start() to block the main-thread, or its the last statement of the main() function?
5. Is creating 1000 messages all with the SAME session-id (so only one session) and receive using the above code sufficient to repro?
5. In addition to the above Processor configurations, code and number of messages, what exact value for "reactor.schedulers.defaultBoundedElasticSize" trigerred the ""hang"" (you can Print this system variable and check if you're choosing default)

**Comment by Maxeh** [comment]
Hi @anuchandy , yes your setup is correct. 
1. There is just one log in this method and the complete() call on the message
2. There is just a log in there
3. We start inside SpringBoot context, but it should not make a difference I guess
4. Receiver is started inside SpringBoot ApplicationListener, start() is the last call, as the receiver is just started and will run on a different thread it should not make a difference right?
5. Yes it is sufficient, in my case exactly 12 messages are received, then the application hangs for sessionIdleTimeout, there a lots of those logs: {"az.sdk.message":"Starting lock renewal.","isSession":true,"lockToken":"1"}; afterwards we see some InterruptedExceptions and then again 12 messages are received and it hangs again for sessionIdleTimeout seconds and so on
7. In the case of sessionProcessor there is no difference in the config of defaultBoundedElasticSize, same behavior for 50 or 100

So to sum it up, prefetch is no usable with sessions. But as I described when NOT using sessions there are also bugs with concurrency and then there is even no real recovery when it hangs and the application needs to be restarted. But then again, when considering prefetchCount + maxConcurrentCalls <= reactor.schedulers.defaultBoundedElasticSize everything seems to work fine - this is how we currently configured it in our system, everything else is not working properly.

**Comment by Maxeh** [comment]
And here some information about the configuration without session (this is our current configuration):
```
final ServiceBusProcessorClient receiver = new ServiceBusClientBuilder()
          .credential(namespace, new ClientSecretCredentialBuilder()
                .tenantId(tenantId)
                .clientId(clientId)
                .clientSecret(clientSecret)
                .build())
          .processor()
          .disableAutoComplete()
          .receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
          .topicName(topic)
          .subscriptionName(subscription)
          .prefetchCount(5)
          .maxConcurrentCalls(25)
          .processError(this::processError)
          .processMessage(this::processMessage)
          .buildProcessorClient();
```

When defaultBoundedElasticSize is 30 everything works fine. When defaultBoundedElasticSize is set to 25 then the application hangs at some point, but it might take over 1000 messages to process until it hangs, but then everything is broken and nothing is working anymore. I have done my tests with at least 5k messages in the queue.

**Comment by anuchandy** [comment]
Hello @Maxeh,

Thanks for the exact setup. I’m able to reproduce the problem you’re experiencing. As you identified, if the pool size is lower than the concurrency and prefetch, this issue of InterruptedException followed by hang happens. This is documented [here ](https://github.com/Azure/azure-sdk-for-java/blob/8b6f7e4b7bde29f5d0a81eb2415beddc9b682db6/sdk/servicebus/azure-messaging-servicebus/TROUBLESHOOTING.md?plain=1#L26) in troubleshooting guidelines.

The on-going work related to Prefetch reliability addresses this problem completely (so no more interruption/hang or need for pool sizing).  I’ve tested your use case using the working PR https://github.com/Azure/azure-sdk-for-java/pull/34854 and it works as expected. Additionally, this work should also addresses the slowness you experienced. I’ll keep this issue open so you / others can track 34854.

**Comment by Maxeh** [comment]
Thanks for the update, good to hear that you are already working on a fix!

**Comment by anuchandy** [comment]
Resolving this. Refer https://learn.microsoft.com/en-us/azure/developer/java/sdk/troubleshooting-messaging-service-bus-overview#upgrade-to-715x for details about version 7.15.x and how to opting-in new session rework.
